packages/cli-kit/src/log.ts

```diff
@@ -0,0 +1,175 @@
+    }
+    callback()
+  }
+
+  _flush(callback: TransformCallback): void {
+    this.push(this.linesToRetain.join(EOL))
+    callback()
+  }
+
+  // Lines retained length average is used so the number of lines depends on the length of them
+  calculateNumLinesToRetain() {
+    return Math.floor(
+      this.maxFileSize /
+        (this.linesToRetain.map((line) => line.length).reduce((l1, l2) => l1 + l2, 0) / this.linesToRetain.length),
+    )
+  }
+
+  completeLastLine(tokens: string[]) {
+    if (this.lastLineCompleted) {
+      return
+    }
+
+    const remainingToken = tokens.shift() ?? ''
+    const incompleteToken = this.linesToRetain[this.linesToRetain.length - 1] ?? ''
+    this.linesToRetain[this.linesToRetain.length - 1] = incompleteToken.concat(remainingToken)
+  }
+}
+
+export async function initiateLogging(options: {logDir?: string; override?: boolean} = {}) {
+  if (isUnitTest()) return
+  commandUuid = generateRandomUUID()
+  logFilePath = getLogFilePath(options)
+  await truncateLogs(logFilePath)
+  logFileStream = createWriteStream(logFilePath, {flags: 'a'})
+}
+
+// DO NOT USE THIS FUNCTION DIRECTLY under normal circumstances.
+// It is exported purely for use in cases where output is already being logged
+// to the terminal but is not reflected in the logfile, e.g. Listr output.
+export function logToFile(message: string, logLevel: string): void {
+  // If file logging hasn't been initiated, skip it
+  if (!logFileExists()) return
+  const timestamp = new Date().toISOString()
+  const logContents = `[${timestamp} ${commandUuid} ${logLevel}]: ${message}\n`
+  logFileStream.write(logContents)
+}
+
+export async function pageLogs({lastCommand}: {lastCommand: boolean}) {
+  const logDir = constants.paths.directories.cache.path()
+  const logFile = pathJoin(logDir, logFileName)
+  // Ensure file exists in case they deleted it or something
+  fileTouchSync(logFile)
+  if (lastCommand) {
+    printLastCommand(logFile)
+  } else {
+    await page(logFile)
+  }
+}
+
+function getLogFilePath(options: {logDir?: string; override?: boolean} = {}) {
+  if (!logFilePath || options.override) {
+    const logDir = options.logDir || constants.paths.directories.cache.path()
+    fileMkdirSync(logDir)
+    logFilePath = pathJoin(logDir, logFileName)
+    fileTouchSync(logFilePath)
+  }
+
+  return logFilePath
+}
+
+// Shaves off older log lines if logs are over maxLogFileSize long.
+async function truncateLogs(logFile: string) {
+  if (truncating || fileSizeSync(logFile) < maxLogFileSize) {
+    return
+  }
+  const tmpLogFile = logFile.concat('.tmp')
+  truncating = true
+  const truncateLines = new LinesTruncatorTransformer(maxLogFileSize)
+  await pipeline(createReadStream(logFile), truncateLines, createWriteStream(tmpLogFile))
+  await pipeline(createReadStream(tmpLogFile), createWriteStream(logFile)).then(() => {
+    unlinkSync(tmpLogFile)
+    truncating = false
+  })
```

Comment:
We try to avoid mixing `await` and `then`, could this be?:
```ts
await pipeline(createReadStream(tmpLogFile), createWriteStream(logFile))
await unlinkSync(tmpLogFile)
truncating = false
```

Reply:
I have little lint problem with the using of `truncating` global variable. I will add an odd workaround to avoid disable the rule